### ISSUE TITLE
Misplaced dot results in suboptimal error messages

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#189]
For example, when parsing

    method test() {
      if (true) {
        . // misplaced dot
      }
    }

the parser reports `Parse error: End` for line 1 (`method test`).

In the attached file `linked-list-predicates-really-cutdown.vpr` (provided by Peter), there are three misplaced dots (`...`) in line 39 and the error message is `Parser error: "}"` for line 37, but there is no `}` close to that line.

I am a bit sceptical that we can do much about this issue without using a different parsing approach, hence the minor priority.

[Issue created by mschwerhoff: 2016-12-14]
[Last updated on bitbucket: 2018-10-17]

[Comment created by fabiopakk: 2018-10-17]
Fixed issue in branch fabiopakk_issue_183_misplaced_dot and merged into default branch.

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2017-05-15 14:25

We should definitely do this.

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2017-05-15 14:31

The following currently fails in Carbon. If we generated triggers internally for the existential (or supported them at the source level) then the proper treatment of heap-dependent triggers should take care of things. By the time we reach the Z3 level, there is no obvious trigger to use except the the function application depending on the (already modified) heap.

```
#!scala
domain Array {
  function loc(a: Array, i: Int): Ref
  function len(a: Array): Int
  function first(r: Ref): Array
  function second(r: Ref): Int

  axiom all_diff {
    forall a: Array, i: Int :: {loc(a, i)}
      first(loc(a, i)) == a && second(loc(a, i)) == i
  }

  axiom length_nonneg {
    forall a: Array :: len(a) >= 0
  }
}

field val : Int


function dummy(j:Int) : Bool
{
  true
}


method qpTest(a: Array, j:Int)
  requires forall i:Int :: {dummy(i)} -2 <= i && i< 8 ==> acc(loc(a,i+2).val)
  requires 0 <= j && j < 10
{
  assume dummy(j-2)
  loc(a,j).val := 5
  assert exists k:Int :: dummy(k-2) && 0 <= k // fails - the "dummy" application is for a different heap
}

```

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2018-10-10 11:33

NOTE: this will entail the following changes:
1. modifying the AST (also ParseAST, I guess)
2. updating any related utility code (e.g. transformer code; most-likely only the `subnodes` definition in the `Nodes` helper object needs changing for that particular purpose?)
3. updating the trigger-generation code to work for existentials. I assume this should be a very minor change (if any) that's needed.
4. updating the back-ends to call the trigger-generation code for existentials as well as foralls, as well as to translate the triggers for existentials down to SMT / Boogie.

**Comment by fpoli** [comment]
> **@meilers** commented on 2019-07-24 12:52

Implemented in Silver commit 5b735cc (and simultaneous commits in Silicon and Carbon).

Alex' example still doesn't verify without manual triggers, and I stared at it a bit and couldn't figure out what trigger on the existential could possibly make it work, but I added a different test case instead.
